src/Sentry/Internal/StructuredLogBatchBuffer.cs

```diff
@@ -0,0 +1,304 @@
+    private SentryLog[] ToArrayAndClear(int length)
+    {
+        Debug.Assert(_addLock.IsSet);
+
+        var array = ToArray(length);
+        Clear(length);
+        return array;
+    }
+
+    private SentryLog[] ToArray(int length)
+    {
+        if (length == 0)
+        {
+            return Array.Empty<SentryLog>();
+        }
+
+        var array = new SentryLog[length];
+        Array.Copy(_array, array, length);
+        return array;
+    }
+
+    private void Clear(int length)
+    {
+        if (length == 0)
+        {
+            return;
+        }
+
+        _additions = 0;
+        Array.Clear(_array, 0, length);
+    }
+
+    private void EnableTimer()
+    {
+        var updated = _timer.Change(_timeout, Timeout.InfiniteTimeSpan);
+        Debug.Assert(updated, "Timer was not successfully enabled.");
+    }
+
+    private void DisableTimer()
+    {
+        var updated = _timer.Change(Timeout.Infinite, Timeout.Infinite);
+        Debug.Assert(updated, "Timer was not successfully disabled.");
+    }
+
+    /// <inheritdoc />
+    public void Dispose()
+    {
+        _timer.Dispose();
+        _addLock.Dispose();
+    }
+
+    private static void ThrowIfLessThanTwo(int value, string paramName)
+    {
+        if (value < 2)
+        {
+            ThrowLessThanTwo(value, paramName);
+        }
+
+        static void ThrowLessThanTwo(int value, string paramName)
+        {
+            throw new ArgumentOutOfRangeException(paramName, value, "Argument must be at least two.");
+        }
+    }
+
+    private static void ThrowIfNegativeOrZero(TimeSpan value, string paramName)
+    {
+        if (value <= TimeSpan.Zero && value != Timeout.InfiniteTimeSpan)
+        {
+            ThrowNegativeOrZero(value, paramName);
+        }
+
+        static void ThrowNegativeOrZero(TimeSpan value, string paramName)
+        {
+            throw new ArgumentOutOfRangeException(paramName, value, "Argument must be a non-negative and non-zero value.");
+        }
+    }
+
+    /// <summary>
+    /// A scope than ensures only a single <see cref="Flush"/> operation is in progress,
+    /// and blocks the calling thread until all <see cref="Add"/> operations have finished.
+    /// When <see cref="IsEntered"/> is <see langword="true"/>, no more <see cref="Add"/> can be started,
+    /// which will then return <see cref="BatchBufferAddStatus.IgnoredIsFlushing"/> immediately.
+    /// </summary>
+    /// <remarks>
+    /// Only <see cref="Flush"/> when scope <see cref="IsEntered"/>.
+    /// </remarks>
+    internal ref struct FlushScope : IDisposable
+    {
+        private StructuredLogBatchBuffer? _lockObj;
+        private ScopedCountdownLock.LockScope _scope;
+
+        internal FlushScope(StructuredLogBatchBuffer lockObj, ScopedCountdownLock.LockScope scope)
+        {
+            _lockObj = lockObj;
+            _scope = scope;
+        }
+
+        internal bool IsEntered => _scope.IsEntered;
```

Comment:
If the `FlushScope` gets created without any scope like here:
https://github.com/getsentry/sentry-dotnet/blob/f28cc6d7750e2b5fea81eaf78c1d0a23033ed2c1/src/Sentry/Internal/StructuredLogBatchBuffer.cs#L129

Then wouldn't this result in a `NullReferenceException`?

It seems like implicitly if there is a `_lockObj` then there should be a `_scope` and `_scope.IsEntered` should be `true`. Possibly this could be an assertion in the constructor above then (make sure `_scope.IsEntered == true`)? Since otherwise the _scope.Wait() here might wait indefinitely:
https://github.com/getsentry/sentry-dotnet/blob/f28cc6d7750e2b5fea81eaf78c1d0a23033ed2c1/src/Sentry/Internal/StructuredLogBatchBuffer.cs#L274

Reply:
Correct.
Great observation!
Adding a `Debug.Assert`.